### PR TITLE
Is not empty should be false when a node is missing

### PR DESCRIPTION
**Pull request by ashleyfrieze**
Possible fix for #53 

@lmartelli  - please review and let me know what you think?

### CODE REVIEW

File: src/main/java/uk/org/webcompere/modelassert/json/dsl/JsonNodeAssertDsl.java [hunk 1]
```diff
@@ -161,6 +161,6 @@ default JsonAssertDslBuilders.At<A> at(String path) {
      * @return the {@link CoreJsonAssertion} for fluent assertions, with this condition added
      */
     default A isNotEmpty() {
-        return satisfies(not(new IsEmpty()));
+        return satisfies(not(MissingCondition.getInstance()).and(not(new IsEmpty())));
```
Comment (lmartelli) [inline src/main/java/uk/org/webcompere/modelassert/json/dsl/JsonNodeAssertDsl.java:164]: It does the job for missing nodes, but what about `null` values ? It feels wrong that `assertJson("{\"prop\": null}").at("/prop").isNotEmpty();` passes.
```suggestion
        return satisfies(not(MissingCondition.getInstance())..and(not(NullCondition.getInstance())).and(not(new IsEmpty())));
```

File: src/test/java/uk/org/webcompere/modelassert/json/condition/IsEmptyTest.java [hunk 2]
```diff
@@ -7,11 +7,18 @@
...
     }
 
+    @Test
+    void isNotEmptyOnMissingIsFalse() {
+        assertAllWays("{\"count\":0}", "{}",
+            assertion -> assertion.at("/count").isNotEmpty());
+    }
+
+
     @Test
```
Comment (lmartelli) [inline src/test/java/uk/org/webcompere/modelassert/json/condition/IsEmptyTest.java:22]: @Test
    void isNotEmptyOnNullIsFalse() {
        assertAllWays("{\"count\":0}", "{\"count\": null}",
            assertion -> assertion.at("/count").isNotEmpty());
    }

**Comment by ashleyfrieze** [comment]
@lmartelli  - I think you're right. I've pushed another extension to the code. See what you think.

**Comment by lmartelli** [comment]
Looks good to me. But I wonder if it's a good think to allow checking for emptyness for all value types. It does not really mean anything for booleans or integers.

**Comment by ashleyfrieze** [comment]
@lmartelli - given that everything in JSON is an `any`, when we're asking `isEmpty` or `isNotEmpty` on a node of `any` type, it's on us to determine if that's meaningful.

I'd be happy that `isEmpty` is true for things where _emptiness_ is a thing - e.g. strings, arrays and objects, and that `isNotEmpty` is both a truthy AND emptiness checker.

Realistically, it's on the user not to use `empty` to check things where emptiness has no meaning.

Have you a counter proposal? By the way, I'm appreciating having someone to discuss these ideas with and shape the tool.

**Comment by lmartelli** [comment]
I encountered the issue because I used `isNotEmpty()` on an expected string. In the end, I used `isNotEmptyText()` which was better suited to my needs.

I am not very comfortable with `isEmpty`/`isNotEmpty` being both a truthy and emptiness checker. Better to be explicit about what you expect. I would deprecate `isEmpty()` and provide `isEmptyArray()` and `isEmptyObject()`.

**Comment by ashleyfrieze** [comment]
@lmartelli I took the spirit of your point and I've implemented something deeper. I also found out that the README recommended the use of more assertions than just `isEmpty`.

It recommended:

```java
assertJson(foo)
   .at("/path").isText()
   .at("/path").isEmpty();
```

or

```java
assertJson(foo)
   .at("/path").isText()
   .at("/path").isNotEmpty();
```

I've now made it so that `.text()`, `.array()` and `.object()` - which specialise the DSL down to things relevant to those types - also assert that you're in the right type.

This means that the preferred route is:

```java
assertJson(foo)
   .at("/path").text().isNotEmpty();
```

And the `isNotEmpty` here is not doing any special check other than `!empty`.

But I've left the shortcuts in place, with a deprecation on `isNotEmpty` when done generically.

I think this allows the shorthand, which is convenient, but also more heavily recommends a stronger-typed approach, which is clearer.

I don't like `isEmptyText` as much.

I need to do this for Boolean and Numeric node DSLs too... but give me early feedback.

**Review by lmartelli — APPROVED** [review]

**Comment by ashleyfrieze** [comment]
@lmartelli - thanks for your input. I've done some further refactoring to make this new approach make sense. This might be a breaking change for some users. I'll get this released in due course.

**Pull request merged**